File: src/lib/export.ts
// Data export utilities — JSON + CSV serialization.
// Pure functions, no I/O. File writing goes through Tauri's fs API.

import type { Summary, Breakdown, Trends, SessionVm } from "./api";

// ── Types ──────────────────────────────────────────────────────────────────

export interface ExportSnapshot {
  period: string;
  summary: Summary;
  breakdown_by_tool: Breakdown;
  breakdown_by_model: Breakdown;
}

export interface ExportPayload {
  generated_at: string;
  app: { name: string; version: string };
  snapshots: ExportSnapshot[];
  daily_trends: Trends;
  sessions: SessionVm[];
}

// ── CSV Helpers ────────────────────────────────────────────────────────────

const BOM = "﻿";

function csvEscape(value: unknown): string {
  const s = value === null || value === undefined ? "" : String(value);
  return /",\n\r]/.test(s) ? `"${s.replace(/"/g, '""')}"` : s;
}

function toCsv(rows: Record<string, unknown>[], columns: string[]): string {
  const header = columns.map(csvEscape).join(",");
  const body = rows.map((row) =>
    columns.map((col) => csvEscape(row[col])).join(",")
  );
  return BOM + [header, ...body].join("\r\n") + "\r\n";
}

// ── Snapshot CSV ───────────────────────────────────────────────────────────

const SNAPSHOT_COLUMNS = [
  "period",
  "dimension",
  "name",
  "tokens",
  "cost_usd",
  "messages",
  "input",
  "output",
  "cache_read",
  "cache_write",
];

export function renderSnapshotCsv(snapshots: ExportSnapshot[]): string {
  const rows: Record<string, unknown>[] = [];
  for (const snap of snapshots) {
    // Tool breakdown
    for (const entry of snap.breakdown_by_tool.entries) {
      rows.push({
        period: snap.period,
        dimension: "tool",
        name: entry.key,
        tokens: entry.tokens,
        cost_usd: entry.cost_usd,
        messages: entry.messages,
        input: entry.input,
        output: entry.output,
        cache_read: entry.cache_read,
        cache_write: entry.cache_write,
      });
    }
    // Model breakdown
    for (const entry of snap.breakdown_by_model.entries) {
      rows.push({
        period: snap.period,
        dimension: "model",
        name: entry.key,
        tokens: entry.tokens,
        cost_usd: entry.cost_usd,
        messages: entry.messages,
        input: entry.input,
        output: entry.output,
        cache_read: entry.cache_read,
        cache_write: entry.cache_write,
      });
    }
  }
  return toCsv(rows, SNAPSHOT_COLUMNS);
}

// ── Daily Trends CSV ───────────────────────────────────────────────────────

const DAILY_COLUMNS = ["date", "tokens", "cost_usd", "messages"];

export function renderDailyCsv(trends: Trends): string {
  const rows = trends.points.map((p) => ({
    date: p.date,
    tokens: p.tokens,
    cost_usd: p.cost_usd,
    messages: p.messages,
  }));
  return toCsv(rows, DAILY_COLUMNS);
}

// ── Sessions CSV ───────────────────────────────────────────────────────────

const SESSION_COLUMNS = [
  "tool",
  "session_id",
  "tokens",
  "cost_usd",
  "messages",
  "rounds",
  "model_count",
  "models",
  "last_used_at",
  "project_name",
];

export function renderSessionsCsv(sessions: SessionVm[]): string {
  const rows = sessions.map((s) => ({
    tool: s.tool,
    session_id: s.session_id,
    tokens: s.tokens,
    cost_usd: s.cost_usd,
    messages: s.messages,
    rounds: s.rounds,
    model_count: s.model_count,
    models: s.models,
    last_used_at: s.last_used_at ?? "",
    project_name: s.project_name ?? "",
  }));
  return toCsv(rows, SESSION_COLUMNS);
}

// ── Full JSON Export ───────────────────────────────────────────────────────

export function renderExportJson(payload: ExportPayload): string {
  return JSON.stringify(payload, null, 2) + "\n";
}

File: src-tauri/src/commands/export.rs
//! Export commands — JSON + CSV serialization of usage data.

use serde::Serialize;
use tauri::State;

use crate::commands::{db, parse_period, today};
use crate::query::{self, breakdown::Breakdown, sessions::SessionVm, trends::Trends, Dimension};
use crate::state::AppState;

#[derive(Serialize)]
struct ExportSnapshot {
    period: String,
    summary: query::summary::Summary,
    breakdown_by_tool: Breakdown,
    breakdown_by_model: Breakdown,
}

#[derive(Serialize)]
struct ExportPayload {
    generated_at: String,
    app: serde_json::Value,
    snapshots: Vec<ExportSnapshot>,
    daily_trends: Trends,
    sessions: Vec<SessionVm>,
}

#[tauri::command]
pub fn export_json(state: State<AppState>) -> Result<String, String> {
    let conn = db(&state);
    let today = today();

    let periods = ["day", "month", "total"];
    let mut snapshots = Vec::new();

    for period_str in &periods {
        let p = parse_period(period_str);
        let range = query::range_for_period(p, &today);
        let summary = query::summary::query(&conn, &range).map_err(|e| e.to_string())?;
        let tool_breakdown =
            query::breakdown::query(&conn, &range, Dimension::Tool).map_err(|e| e.to_string())?;
        let model_breakdown =
            query::breakdown::query(&conn, &range, Dimension::Model).map_err(|e| e.to_string())?;

        snapshots.push(ExportSnapshot {
            period: period_str.to_string(),
            summary,
            breakdown_by_tool: tool_breakdown,
            breakdown_by_model: model_breakdown,
        });
    }

    let total_range = query::range_for_period(query::Period::Total, &today);
    let daily_trends = query::trends::query(&conn, &total_range).map_err(|e| e.to_string())?;

    // Get Claude projects dir for session queries
    let claude_projects_dir = dirs::home_dir().map(|h| h.join(".claude").join("projects"));
    let sessions = query::sessions::query(&conn, claude_projects_dir.as_deref(), Some(500))
        .map_err(|e| e.to_string())?;

    let payload = ExportPayload {
        generated_at: chrono::Utc::now().to_rfc3339(),
        app: serde_json::json!({
            "name": "token-usage",
            "version": env!("CARGO_PKG_VERSION"),
        }),
        snapshots,
        daily_trends,
        sessions,
    };

    serde_json::to_string_pretty(&payload).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn export_csv(state: State<AppState>) -> Result<String, String> {
    let conn = db(&state);
    let today = today();

    let periods = ["day", "month", "total"];
    let mut rows: Vec<Vec<String>> = Vec::new();
    rows.push(vec![
        "period".into(),
        "dimension".into(),
        "name".into(),
        "tokens".into(),
        "cost_usd".into(),
        "messages".into(),
        "input".into(),
        "output".into(),
        "cache_read".into(),
        "cache_write".into(),
    ]);

    for period_str in &periods {
        let p = parse_period(period_str);
        let range = query::range_for_period(p, &today);

        for (dim_label, dim) in [("tool", Dimension::Tool), ("model", Dimension::Model)] {
            let breakdown =
                query::breakdown::query(&conn, &range, dim).map_err(|e| e.to_string())?;
            for entry in &breakdown.entries {
                rows.push(vec![
                    period_str.to_string(),
                    dim_label.to_string(),
                    csv_escape(&entry.key),
                    entry.tokens.to_string(),
                    format!("{:.6}", entry.cost_usd),
                    entry.messages.to_string(),
                    entry.input.to_string(),
                    entry.output.to_string(),
                    entry.cache_read.to_string(),
                    entry.cache_write.to_string(),
                ]);
            }
        }
    }

    Ok(rows
        .iter()
        .map(|r| r.join(","))
        .collect::<Vec<_>>()
        .join("\r\n"))
}

fn csv_escape(value: &str) -> String {
    if value.contains(',') || value.contains('"') || value.contains('\n') {
        format!("\"{}\"", value.replace('"', "\"\""))
    } else {
        value.to_string()
    }
}

#[tauri::command]
pub fn copy_to_clipboard(text: String) -> Result<(), String> {
    // Use arboard for clipboard access
    let mut clipboard = arboard::Clipboard::new().map_err(|e| e.to_string())?;
    clipboard.set_text(text).map_err(|e| e.to_string())?;
    Ok(())
}
